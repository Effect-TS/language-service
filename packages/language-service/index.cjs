// eslint-disable-next-line @typescript-eslint/consistent-type-imports, no-undef
module.exports = function init(modules) {
  function create(info) {
    // start with regular language service, then load custom once async module has loaded
    let languageService = info.languageService
    import("./build/cjs/index.js").then(v => languageService = v.default(modules).create(info))

    // create the proxy
    const proxy = Object.create(null)
    for (const k of Object.keys(info.languageService)) {
      proxy[k] = (...args) => languageService[k].apply(languageService, args)
    }

    return proxy
  }

  return { create }
}
