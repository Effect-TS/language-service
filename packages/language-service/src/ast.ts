import { Tag } from "@tsplus/stdlib/service/Tag"
import type ts from "typescript/lib/tsserverlibrary"

declare module "typescript/lib/tsserverlibrary" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any
  export type ChangeNodeOptions = any

  export namespace textChanges {
    export class ChangeTracker {
      static with(context: ts.TextChangesContext, cb: (tracker: ChangeTracker) => void): ts.FileTextChanges[]
      replaceNode(sourceFile: ts.SourceFile, oldNode: ts.Node, newNode: ts.Node, options?: ts.ChangeNodeOptions): void
    }
    export function applyChanges(text: string, changes: readonly ts.TextChange[]): string
  }

  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFileLike,
    startNode: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFile,
    startNode?: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined

  export function isMemberName(node: ts.Node): node is ts.MemberName
  export function isKeyword(token: ts.SyntaxKind): token is ts.KeywordSyntaxKind
}

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export type TypeScriptApi = typeof import("typescript/lib/tsserverlibrary")
export const TypeScriptApi = Tag<TypeScriptApi>()
export const LanguageServiceApi = Tag<ts.LanguageService>()
export const ChangeTrackerApi = Tag<ts.textChanges.ChangeTracker>()

export class NoTypeScriptProgramError {
  readonly _tag = "NoTypeScriptProgramError"
}

export function getProgram() {
  return Effect.serviceWithEffect(
    LanguageServiceApi,
    languageService => Effect.sync(() => languageService.getProgram())
  )
    .filterOrFail((program): program is ts.Program => !!program, () => new NoTypeScriptProgramError())
}

export class NoSuchSourceFile {
  readonly _tag = "NoSuchSourceFile"
  constructor(
    readonly fileName: string
  ) {}
}

export function getSourceFile(fileName: string) {
  return getProgram().map(program => program.getSourceFile(fileName))
    .filterOrFail((sourceFile): sourceFile is ts.SourceFile => !!sourceFile, () => new NoSuchSourceFile(fileName))
}

export function hasModifier(node: ts.Declaration, kind: ts.ModifierFlags) {
  return Effect.serviceWith(TypeScriptApi, ts => !!(ts.getCombinedModifierFlags(node) & kind))
}

function getChildNodesContainingRange(
  parent: ts.Node,
  textRange: ts.TextRange
) {
  return Do($ => {
    const ts = $(Effect.service(TypeScriptApi))
    let result: Chunk<ts.Node> = Chunk.empty()

    ts.forEachChild(parent, node => {
      if (node.end < textRange.pos) {
        return
      } else if (node.pos > textRange.end) {
        return
      } else {
        result = result.append(node)
      }
    })

    return result
  })
}

/**
 * Gets the closest node that contains given TextRange
 */
export function getNodesContainingRange(
  sourceFile: ts.SourceFile,
  textRange: ts.TextRange
) {
  return Do(($) => {
    let result: Chunk<ts.Node> = Chunk.empty()

    $(
      Effect.iterate([sourceFile] as ts.Node[], (_) => _.length > 0)(toProcess => {
        const node = toProcess.shift()!
        return getChildNodesContainingRange(node, textRange).flatMap(_ =>
          Effect.sync(() => {
            result = result.concat(_)
            return toProcess.concat(..._.toArray)
          })
        )
      })
    )

    return Chunk.from(result.reverse)
  })
}

/**
 * Ensures value is a text range
 */
export function toTextRange(positionOrRange: number | ts.TextRange): ts.TextRange {
  return typeof positionOrRange === "number" ? { end: positionOrRange, pos: positionOrRange } : positionOrRange
}

export function getHumanReadableName(sourceFile: ts.SourceFile, node: ts.Node) {
  const text = node.getText(sourceFile)
  return text.length > 10 ? text.substring(0, 10) + "..." : text
}

export function collectAll<R, E>(rootNode: ts.Node, test: (node: ts.Node) => Effect<R, E, boolean>) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const env = $(Effect.environment<R>())
    let result = Chunk.empty<ts.Node>()

    function visitor(node: ts.Node) {
      if (test(node).provideEnvironment(env).unsafeRunSync()) result = result.append(node)
      ts.visitEachChild(node, visitor, ts.nullTransformationContext)

      return node
    }

    ts.visitNode(rootNode, visitor)

    return result
  })
}

export function getRelevantTokens(
  position: number,
  sourceFile: ts.SourceFile
) {
  return Do(($) => {
    const ts = $(Effect.service(TypeScriptApi))
    const previousToken = ts.findPrecedingToken(position, sourceFile)
    if (
      previousToken && position <= previousToken.end &&
      (ts.isMemberName(previousToken) || ts.isKeyword(previousToken.kind))
    ) {
      const contextToken = ts.findPrecedingToken(previousToken.getFullStart(), sourceFile, /*startNode*/ undefined)! // TODO: GH#18217
      return { contextToken, previousToken }
    }
    return { contextToken: Maybe.fromNullable(previousToken), previousToken: Maybe.fromNullable(previousToken) }
  })
}
