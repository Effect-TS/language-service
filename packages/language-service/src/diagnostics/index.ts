import removeCurryArrow from "@effect/language-service/diagnostics/removeCurryArrow"

export default { removeCurryArrow }
