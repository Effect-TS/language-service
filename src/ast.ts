import * as T from "@effect/core/io/Effect"
import * as Ch from "@tsplus/stdlib/collections/Chunk"
import { pipe } from "@tsplus/stdlib/data/Function"
import * as O from "@tsplus/stdlib/data/Maybe"
import { Tag } from "@tsplus/stdlib/service/Tag"
import type ts from "typescript/lib/tsserverlibrary"

declare module "typescript/lib/tsserverlibrary" {
  const nullTransformationContext: ts.TransformationContext

  export namespace formatting {
    interface FormattingHost {
      getNewLine?(): string
    }

    export interface FormatContext {
      readonly options: ts.FormatCodeSettings
      readonly getRules: unknown
    }

    function getFormatContext(options: ts.FormatCodeSettings, host: FormattingHost): FormatContext
  }

  export type TextChangesContext = any
  export type ChangeNodeOptions = any

  export namespace textChanges {
    export enum LeadingTriviaOption {
      /** Exclude all leading trivia (use getStart()) */
      Exclude = 0,
      /** Include leading trivia and,
       * if there are no line breaks between the node and the previous token,
       * include all trivia between the node and the previous token
       */
      IncludeAll = 1,
      /**
       * Include attached JSDoc comments
       */
      JSDoc = 2,
      /**
       * Only delete trivia on the same line as getStart().
       * Used to avoid deleting leading comments
       */
      StartLine = 3
    }
    export enum TrailingTriviaOption {
      /** Exclude all trailing trivia (use getEnd()) */
      Exclude = 0,
      /** Doesn't include whitespace, but does strip comments */
      ExcludeWhitespace = 1,
      /** Include trailing trivia */
      Include = 2
    }
    export interface ConfigurableStart {
      leadingTriviaOption?: LeadingTriviaOption
    }
    export interface ConfigurableEnd {
      trailingTriviaOption?: TrailingTriviaOption
    }
    export interface InsertNodeOptions {
      /**
       * Text to be inserted before the new node
       */
      prefix?: string
      /**
       * Text to be inserted after the new node
       */
      suffix?: string
      /**
       * Text of inserted node will be formatted with this indentation, otherwise indentation will be inferred from the old node
       */
      indentation?: number
      /**
       * Text of inserted node will be formatted with this delta, otherwise delta will be inferred from the new node kind
       */
      delta?: number
    }
    export interface ConfigurableStartEnd extends ConfigurableStart, ConfigurableEnd {
    }
    export class ChangeTracker {
      static with(context: ts.TextChangesContext, cb: (tracker: ChangeTracker) => void): Array<ts.FileTextChanges>
      delete(sourceFile: ts.SourceFile, node: ts.Node | ts.NodeArray<ts.TypeParameterDeclaration>): void
      deleteRange(sourceFile: ts.SourceFile, range: ts.TextRange): void
      replaceNode(sourceFile: ts.SourceFile, oldNode: ts.Node, newNode: ts.Node, options?: ts.ChangeNodeOptions): void
      insertNodeAt(
        sourceFile: ts.SourceFile,
        pos: number,
        newNode: ts.Node,
        options?: ts.textChanges.InsertNodeOptions
      ): void
      insertNodeBefore(
        sourceFile: ts.SourceFile,
        before: ts.Node,
        newNode: ts.Node,
        blankLineBetween?: boolean,
        options?: ConfigurableStartEnd
      ): void
      insertNodeAfter(sourceFile: ts.SourceFile, after: ts.Node, newNode: ts.Node): void
      insertText(sourceFile: ts.SourceFile, pos: number, text: string): void
    }
    export function applyChanges(text: string, changes: ReadonlyArray<ts.TextChange>): string
  }

  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFileLike,
    startNode: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  export function findPrecedingToken(
    position: number,
    sourceFile: ts.SourceFile,
    startNode?: ts.Node,
    excludeJsdoc?: boolean
  ): ts.Node | undefined
  function findChildOfKind<T extends ts.Node>(n: ts.Node, kind: T["kind"], sourceFile: ts.SourceFileLike): T | undefined

  export function isMemberName(node: ts.Node): node is ts.MemberName
  export function isKeyword(token: ts.SyntaxKind): token is ts.KeywordSyntaxKind
}

// eslint-disable-next-line @typescript-eslint/consistent-type-imports
export type TypeScriptApi = typeof import("typescript/lib/tsserverlibrary")
export const TypeScriptApi = Tag<TypeScriptApi>()
export const TypeScriptProgram = Tag<ts.Program>()
export const ChangeTrackerApi = Tag<ts.textChanges.ChangeTracker>()

export class NoSuchSourceFile {
  readonly _tag = "NoSuchSourceFile"
  constructor(
    readonly fileName: string
  ) {}
}

export function getSourceFile(fileName: string) {
  return pipe(
    T.service(TypeScriptProgram),
    T.map((program) => program.getSourceFile(fileName)),
    T.filterOrFail((sourceFile): sourceFile is ts.SourceFile => !!sourceFile, () => new NoSuchSourceFile(fileName))
  )
}

export function hasModifier(node: ts.Declaration, kind: ts.ModifierFlags) {
  return T.serviceWith(TypeScriptApi, (ts) => !!(ts.getCombinedModifierFlags(node) & kind))
}

/**
 * Gets the closest node that contains given TextRange
 */
export function getNodesContainingRange(
  ts: TypeScriptApi
) {
  return ((sourceFile: ts.SourceFile, textRange: ts.TextRange) => {
    const precedingToken = ts.findPrecedingToken(textRange.pos, sourceFile)
    if (!precedingToken) return Ch.empty()

    let result = Ch.empty<ts.Node>()
    let parent = precedingToken
    while (parent) {
      result = pipe(result, Ch.append(parent))
      parent = parent.parent
    }

    return result
  })
}

/**
 * Ensures value is a text range
 */
export function toTextRange(positionOrRange: number | ts.TextRange): ts.TextRange {
  return typeof positionOrRange === "number" ? { end: positionOrRange, pos: positionOrRange } : positionOrRange
}

export function getHumanReadableName(sourceFile: ts.SourceFile, node: ts.Node) {
  const text = node.getText(sourceFile)
  return text.length > 10 ? text.substring(0, 10) + "..." : text
}

export function collectAll(ts: TypeScriptApi) {
  return <A extends ts.Node>(rootNode: ts.Node, test: (node: ts.Node) => node is A) => {
    let result = Ch.empty<A>()

    function visitor(node: ts.Node) {
      if (test(node)) result = pipe(result, Ch.append(node))
      ts.forEachChild(node, visitor)
    }

    visitor(rootNode)

    return result
  }
}

export function getRelevantTokens(
  ts: TypeScriptApi
) {
  return ((position: number, sourceFile: ts.SourceFile) => {
    const previousToken = ts.findPrecedingToken(position, sourceFile)
    if (
      previousToken && position <= previousToken.end &&
      (ts.isMemberName(previousToken) || ts.isKeyword(previousToken.kind))
    ) {
      const contextToken = ts.findPrecedingToken(previousToken.getFullStart(), sourceFile, /*startNode*/ undefined)! // TODO: GH#18217
      return { contextToken: O.some(contextToken), previousToken: O.some(previousToken) }
    }
    return { contextToken: O.fromNullable(previousToken), previousToken: O.fromNullable(previousToken) }
  })
}

export function isNodeInRange(textRange: ts.TextRange) {
  return (node: ts.Node) => node.pos <= textRange.pos && node.end >= textRange.end
}
