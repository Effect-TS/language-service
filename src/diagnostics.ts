import { duplicatePackage } from "./diagnostics/duplicatePackage.js"
import { effectInVoidSuccess } from "./diagnostics/effectInVoidSuccess.js"
import { floatingEffect } from "./diagnostics/floatingEffect.js"
import { genericEffectServices } from "./diagnostics/genericEffectServices.js"
import { importFromBarrel } from "./diagnostics/importFromBarrel.js"
import { leakingRequirements } from "./diagnostics/leakingRequirements.js"
import { missingEffectContext } from "./diagnostics/missingEffectContext.js"
import { missingEffectError } from "./diagnostics/missingEffectError.js"
import { missingReturnYieldStar } from "./diagnostics/missingReturnYieldStar.js"
import { missingStarInYieldEffectGen } from "./diagnostics/missingStarInYieldEffectGen.js"
import { returnEffectInGen } from "./diagnostics/returnEffectInGen.js"
import { scopeInLayerEffect } from "./diagnostics/scopeInLayerEffect.js"
import { tryCatchInEffectGen } from "./diagnostics/tryCatchInEffectGen.js"
import { unnecessaryEffectGen } from "./diagnostics/unnecessaryEffectGen.js"
import { unnecessaryPipe } from "./diagnostics/unnecessaryPipe.js"

export const diagnostics = [
  duplicatePackage,
  missingEffectContext,
  missingEffectError,
  floatingEffect,
  missingStarInYieldEffectGen,
  unnecessaryEffectGen,
  missingReturnYieldStar,
  leakingRequirements,
  unnecessaryPipe,
  genericEffectServices,
  returnEffectInGen,
  tryCatchInEffectGen,
  importFromBarrel,
  scopeInLayerEffect,
  effectInVoidSuccess
]
