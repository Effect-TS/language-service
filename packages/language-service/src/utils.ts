import * as AST from "@effect/language-service/ast"
import type ts from "typescript/lib/tsserverlibrary"

export function isPipeCall(node: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    if (!ts.isCallExpression(node)) return false
    const expression = node.expression
    if (!ts.isIdentifier(expression)) return false
    if (expression.getText(node.getSourceFile()) !== "pipe") return false
    return true
  })
}

export function asPipeableCallExpression(node: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    // ensure the node is a call expression
    if (!ts.isCallExpression(node)) return Maybe.none
    // with just 1 arg
    if (node.arguments.length !== 1) return Maybe.none
    const arg = node.arguments[0]!
    // ideally T.map(n => n * 2) could be piped to pipe(n => n * 2, T.map)
    // but does not make any sense.
    if (ts.isArrowFunction(arg)) return Maybe.none
    // same goes for identifiers, string literal or numbers
    if (ts.isStringLiteral(arg) || ts.isNumericLiteral(arg) || ts.isIdentifier(arg)) return Maybe.none
    return Maybe.some([node.expression, arg] as const)
  })
}

export function asPipeArguments(initialNode: ts.Node) {
  return Do($ => {
    let result = Chunk.empty<ts.Expression>()
    $(
      Effect.iterate(Maybe.some(initialNode), node => node.isSome())(
        maybeNode =>
          Do($ => {
            if (maybeNode.isNone()) return Maybe.none

            const node = maybeNode.value
            const maybePipeable = $(asPipeableCallExpression(node))
            if (maybePipeable.isNone()) {
              result = result.append(node as ts.Expression)
              return Maybe.none
            }
            const [exp, arg] = maybePipeable.value
            result = result.append(exp)

            return Maybe.some(arg)
          })
      )
    )
    return result.reverse
  })
}

export function isPipeableCallExpression(node: ts.Node) {
  return asPipeableCallExpression(node).map(_ => _.isSome())
}

export function isCurryArrow(arrow: ts.Node) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))
    if (!ts.isArrowFunction(arrow)) return false
    if (arrow.parameters.length !== 1) return false
    const parameter = arrow.parameters[0]!
    const parameterName = parameter.name
    if (!ts.isIdentifier(parameterName)) return false
    const body = arrow.body
    if (!ts.isCallExpression(body)) return false
    const args = body.arguments
    if (args.length !== 1) return false
    const identifier = args[0]!
    if (!ts.isIdentifier(identifier)) return false
    return identifier.text === parameterName.text
  })
}

export function transformAsyncAwaitToEffectGen(
  node: ts.FunctionDeclaration | ts.ArrowFunction | ts.FunctionExpression,
  onAwait: (expression: ts.Expression) => ts.Expression
) {
  return Do($ => {
    const ts = $(Effect.service(AST.TypeScriptApi))

    function visitor(_: ts.Node): ts.Node {
      if (ts.isAwaitExpression(_)) {
        const expression = ts.visitEachChild(_.expression, visitor, ts.nullTransformationContext)

        return ts.factory.createYieldExpression(
          ts.factory.createToken(ts.SyntaxKind.AsteriskToken),
          ts.factory.createCallExpression(ts.factory.createIdentifier("$"), undefined, [onAwait(expression)])
        )
      }
      return ts.visitEachChild(_, visitor, ts.nullTransformationContext)
    }
    const generatorBody = visitor(node.body!)

    const generator = ts.factory.createFunctionExpression(
      [],
      ts.factory.createToken(ts.SyntaxKind.AsteriskToken),
      undefined,
      [],
      [ts.factory.createParameterDeclaration(undefined, undefined, "$")],
      undefined,
      generatorBody as any // NOTE(mattia): intended, to use same routine for both ConciseBody and Body
    )

    const effectGenCallExp = ts.factory.createCallExpression(
      ts.factory.createPropertyAccessExpression(
        ts.factory.createIdentifier("Effect"),
        "gen"
      ),
      undefined,
      [generator as any]
    )

    let currentFlags = ts.getCombinedModifierFlags(node)
    currentFlags &= ~ts.ModifierFlags.Async
    const newModifiers = ts.factory.createModifiersFromModifierFlags(currentFlags)

    if (ts.isArrowFunction(node)) {
      return ts.factory.createArrowFunction(
        newModifiers,
        node.typeParameters,
        node.parameters,
        undefined,
        node.equalsGreaterThanToken,
        effectGenCallExp
      )
    }

    const newBody = ts.factory.createBlock([
      ts.factory.createReturnStatement(effectGenCallExp)
    ])

    if (ts.isFunctionDeclaration(node)) {
      return ts.factory.createFunctionDeclaration(
        newModifiers,
        node.asteriskToken,
        node.name,
        node.typeParameters,
        node.parameters,
        undefined,
        newBody
      )
    }
    return ts.factory.createFunctionExpression(
      newModifiers,
      node.asteriskToken,
      node.name,
      node.typeParameters,
      node.parameters,
      undefined,
      newBody
    )
  })
}
