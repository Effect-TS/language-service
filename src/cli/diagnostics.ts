import * as Command from "@effect/cli/Command"
import * as Options from "@effect/cli/Options"
import * as Path from "@effect/platform/Path"
import { createProjectService } from "@typescript-eslint/project-service"
import * as Array from "effect/Array"
import * as Data from "effect/Data"
import * as Effect from "effect/Effect"
import * as Either from "effect/Either"
import { pipe } from "effect/Function"
import * as Option from "effect/Option"
import * as Predicate from "effect/Predicate"
import type * as ts from "typescript"
import * as LanguageServicePluginOptions from "../core/LanguageServicePluginOptions"
import * as LSP from "../core/LSP"
import * as Nano from "../core/Nano"
import * as TypeCheckerApi from "../core/TypeCheckerApi"
import * as TypeCheckerUtils from "../core/TypeCheckerUtils"
import * as TypeParser from "../core/TypeParser"
import * as TypeScriptApi from "../core/TypeScriptApi"
import * as TypeScriptUtils from "../core/TypeScriptUtils"
import { diagnostics as diagnosticsDefinitions } from "../diagnostics"
import { getTypeScript } from "./utils"

export class NoFilesToCheckError extends Data.TaggedError("NoFilesToCheckError")<{}> {
  get message(): string {
    return "No files to check. Please provide an existing .ts file or a project tsconfig.json"
  }
}

const file = Options.file("file").pipe(
  Options.optional,
  Options.withDescription("The full path of the file to check for diagnostics.")
)

const project = Options.file("project").pipe(
  Options.optional,
  Options.withDescription("The full path of the project tsconfig.json file to check for diagnostics.")
)

const extractEffectLspOptions = (compilerOptions: ts.CompilerOptions) => {
  return (Predicate.hasProperty(compilerOptions, "plugins") && Array.isArray(compilerOptions.plugins)
    ? compilerOptions.plugins
    : []).find((_) => Predicate.hasProperty(_, "name") && _.name === "@effect/language-service")
}

export const diagnostics = Command.make(
  "diagnostics",
  { file, project },
  Effect.fn("diagnostics")(function*({ file, project }) {
    const path = yield* Path.Path
    const tsInstance = yield* getTypeScript
    const filesToCheck = new Set<string>()
    let checkedFilesCount = 0
    let errorsCount = 0
    let warningsCount = 0
    let messagesCount = 0

    const { service } = createProjectService({ options: { loadTypeScriptPlugins: false } })

    if (Option.isSome(file)) {
      filesToCheck.add(path.resolve(file.value))
    }

    if (Option.isSome(project)) {
      let tsconfigToHandle = [project.value ?? ""]
      while (tsconfigToHandle.length > 0) {
        const tsconfigPath = tsconfigToHandle.shift()!
        const tsconfigAbsolutePath = path.resolve(tsconfigPath)
        // process the project references, which may be a folder with a contained tsconfig.json file
        const configFile = tsInstance.readConfigFile(tsconfigAbsolutePath, tsInstance.sys.readFile)
        if (configFile.error && !tsconfigAbsolutePath.endsWith("tsconfig.json")) {
          tsconfigToHandle = [...tsconfigToHandle, path.resolve(tsconfigPath, "tsconfig.json")]
          continue
        }
        // parse the config file
        const parsedConfig = tsInstance.parseJsonConfigFileContent(
          configFile.config,
          tsInstance.sys,
          path.dirname(tsconfigAbsolutePath)
        )
        tsconfigToHandle = [...tsconfigToHandle, ...parsedConfig.projectReferences?.map((_) => _.path) ?? []]
        parsedConfig.fileNames.forEach((_) => filesToCheck.add(_))
      }
    }

    if (filesToCheck.size === 0) {
      return yield* new NoFilesToCheckError()
    }

    for (const filePath of filesToCheck) {
      service.openClientFile(filePath)
      try {
        const scriptInfo = service.getScriptInfo(filePath)
        if (!scriptInfo) continue

        const project = scriptInfo.getDefaultProject()
        const languageService = project.getLanguageService(true)
        const program = languageService.getProgram()
        if (!program) continue
        const sourceFile = program.getSourceFile(filePath)
        if (!sourceFile) continue
        const pluginConfig = extractEffectLspOptions(program.getCompilerOptions())
        if (!pluginConfig) continue

        // run the diagnostics and pipe them into addDiagnostic
        const results = pipe(
          LSP.getSemanticDiagnosticsWithCodeFixes(diagnosticsDefinitions, sourceFile),
          TypeParser.nanoLayer,
          TypeCheckerUtils.nanoLayer,
          TypeScriptUtils.nanoLayer,
          Nano.provideService(TypeCheckerApi.TypeCheckerApi, program.getTypeChecker()),
          Nano.provideService(TypeScriptApi.TypeScriptProgram, program),
          Nano.provideService(TypeScriptApi.TypeScriptApi, tsInstance),
          Nano.provideService(
            LanguageServicePluginOptions.LanguageServicePluginOptions,
            LanguageServicePluginOptions.parse(pluginConfig)
          ),
          Nano.run,
          Either.map((_) => _.diagnostics),
          Either.map(
            Array.map((_) =>
              _.category === tsInstance.DiagnosticCategory.Suggestion
                ? { ..._, category: tsInstance.DiagnosticCategory.Message }
                : _
            )
          ),
          Either.getOrElse(() => [])
        )
        checkedFilesCount++
        errorsCount += results.filter((_) => _.category === tsInstance.DiagnosticCategory.Error).length
        warningsCount += results.filter((_) => _.category === tsInstance.DiagnosticCategory.Warning).length
        messagesCount += results.filter((_) => _.category === tsInstance.DiagnosticCategory.Message).length
        if (results.length === 0) continue
        // pretty print the results
        const formattedResults = tsInstance.formatDiagnosticsWithColorAndContext(results, {
          getCanonicalFileName: (fileName) => path.resolve(fileName),
          getCurrentDirectory: () => path.resolve("."),
          getNewLine: () => "\n"
        })
        console.log(formattedResults)
      } finally {
        service.closeClientFile(filePath)
      }
      yield* Effect.yieldNow()
    }
    console.log(
      `Checked ${checkedFilesCount} files out of ${filesToCheck.size} files. \n${errorsCount} errors, ${warningsCount} warnings and ${messagesCount} messages.`
    )
  })
)
