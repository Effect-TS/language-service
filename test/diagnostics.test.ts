import * as T from "@effect/io/Effect"
import * as AST from "@effect/language-service/ast"
import type { DiagnosticDefinition } from "@effect/language-service/diagnostics/definition"
import diagnostics from "@effect/language-service/diagnostics/index"
import { createMockLanguageServiceHost } from "@effect/language-service/test/utils"
import * as Ch from "@fp-ts/data/Chunk"
import { pipe } from "@fp-ts/data/Function"
import * as fs from "fs"
import * as path from "path"
import ts from "typescript/lib/tsserverlibrary"

export function testDiagnosticOnExample(diagnostic: DiagnosticDefinition, fileName: string) {
  const sourceText = fs.readFileSync(path.join(__dirname, "..", "examples", "diagnostics", fileName))
    .toString("utf8")

  // create the language service
  const languageServiceHost = createMockLanguageServiceHost(fileName, sourceText)
  const languageService = ts.createLanguageService(languageServiceHost, undefined, ts.LanguageServiceMode.Semantic)
  const program = languageService.getProgram()
  if (!program) throw new Error("No typescript program!")
  const sourceFile = program.getSourceFile(fileName)
  if (!sourceFile) throw new Error("No source file " + fileName + " in VFS")

  // ensure there are no errors in TS file
  const diagnostics = languageService.getCompilerOptionsDiagnostics()
    .concat(languageService.getSyntacticDiagnostics(fileName))
    .concat(languageService.getSemanticDiagnostics(fileName)).map((diagnostic) => {
      const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n")
      if (diagnostic.file) {
        const { character, line } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!)
        return `  Error ${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`
      } else {
        return `  Error: ${message}`
      }
    })
  expect(diagnostics).toEqual([])

  // check and assert the diagnostic runs
  const resultMessages = pipe(
    diagnostic
      .apply(sourceFile),
    T.provideService(AST.TypeScriptApi)(ts),
    T.provideService(AST.TypeScriptProgram)(program),
    T.unsafeRunSync
  )

  // create human readable messages
  const humanMessages = pipe(
    resultMessages,
    Ch.map((error) => {
      const start = ts.getLineAndCharacterOfPosition(sourceFile, error.node.pos)
      const end = ts.getLineAndCharacterOfPosition(sourceFile, error.node.end)
      const errorSourceCode = sourceText.substring(error.node.pos, error.node.end)

      return errorSourceCode + "\n" +
        `${start.line + 1}:${start.character} - ${end.line + 1}:${end.character} | ${error.messageText}`
    }),
    Ch.join("\n\n")
  )
  expect(humanMessages).toMatchSnapshot()
}

function testFiles(name: string, refactor: DiagnosticDefinition, fileNames: Array<string>) {
  for (const fileName of fileNames) {
    describe(fileName, () => {
      it(fileName, () => {
        testDiagnosticOnExample(refactor, fileName)
      })
    })
  }
}

Object.keys(diagnostics).map((diagnosticName) =>
  testFiles(diagnosticName, diagnostics[diagnosticName]!, [diagnosticName + ".ts"])
)
