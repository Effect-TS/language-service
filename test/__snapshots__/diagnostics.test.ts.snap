// Vitest Snapshot v1

exports[`ensureGetCallTrace.ts > ensureGetCallTrace.ts 1`] = `
" <A>(value: A) => T.succeed(value)
5:28 - 5:62 | This function should be traced. getCallTrace() should be the first line of the body.

 <A, B>(f: (value: A) => B) => T.map(f)
7:24 - 7:63 | This function should be traced. getCallTrace() should be the first line of the body.

 <A, B>(f: (value: A) => T.Effect<never, never, B>) => T.flatMap(f)
9:21 - 9:88 | This function should be traced. getCallTrace() should be the first line of the body."
`;

exports[`noSyncWithConstant.ts > noSyncWithConstant.ts 1`] = `
" T.sync(() => 42)
3:10 - 3:27 | Value is constant, instead of using sync you could use succeed.

 T.sync(() => true)
4:10 - 4:29 | Value is constant, instead of using sync you could use succeed.

 T.failSync(() => 42)
13:10 - 13:31 | Value is constant, instead of using failSync you could use fail.

 T.dieSync(() => 42)
16:10 - 16:30 | Value is constant, instead of using dieSync you could use die."
`;
